File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* listNode = new ListNode(0);
        ListNode* currNode = listNode;

        while(l1||l2||carry){
            int x = l1 ? l1->val : 0; 
            int y = l2 ? l2->val : 0;

            int sum = x + y + carry;
            carry = sum/10;
            currNode->next = new ListNode(sum%10);
            currNode = currNode->next;
            
            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
        }

        ListNode* result = listNode->next;
        delete listNode;

        return result;
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let carry = 0;
    let currNode = new ListNode(0, null);
    let listNode = currNode;
    while (l1 || l2 || carry) {
        let acc = ((l1?.val ?? 0) + (l2?.val ?? 0) + carry)
        let sum = acc % 10;
        carry = ~~(acc / 10);
        currNode.next = new ListNode(sum);
        currNode = currNode.next

        l1 = l1?.next ?? null;
        l2 = l2?.next ?? null;
    }


    return listNode.next
};